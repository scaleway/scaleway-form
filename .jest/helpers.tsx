import { ThemeProvider } from '@emotion/react'
import makeHelpers from '@scaleway/jest-helpers'
import { theme as lightTheme } from '@scaleway/ui'
import { render } from '@testing-library/react'
import type { ComponentProps, ReactElement, ReactNode } from 'react'
import { Form } from '../src/components'
import { mockErrors } from '../src/mocks'

type WrapperProps = {
  theme?: typeof lightTheme
  children: ReactNode
}

const Wrapper = ({ theme = lightTheme, children }: WrapperProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotWithPortal,
} = makeHelpers(Wrapper, { classNameReplacer: className => className })

export const shouldMatchEmotionSnapshotFormWrapper = (
  children: ReactNode,
  options?: Parameters<typeof shouldMatchEmotionSnapshot>[1],
  formOptions?: Partial<ComponentProps<typeof Form>>,
) => {
  const { initialValues } = formOptions ?? {}

  return shouldMatchEmotionSnapshot(
    <Form
      onRawSubmit={() => {}}
      errors={mockErrors}
      initialValues={initialValues}
    >
      {() => children}
    </Form>,
    options,
  )
}

export const renderWithWrapper = (children: ReactElement) =>
  render(children, {
    wrapper: () => (
      <ThemeProvider theme={lightTheme}>
        <Form onRawSubmit={() => {}} errors={mockErrors}>
          {children}
        </Form>
      </ThemeProvider>
    ),
  })

export const mockRandom = () =>
  jest.spyOn(global.Math, 'random').mockReturnValue(0.4155913669444804)

export const restoreRandom = () =>
  jest.spyOn(global.Math, 'random').mockRestore()
