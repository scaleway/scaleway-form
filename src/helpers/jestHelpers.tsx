import { ThemeProvider } from '@emotion/react'
import makeHelpers from '@scaleway/jest-helpers'
import { theme as scwuiTheme } from '@scaleway/ui'
import { RenderResult, render } from '@testing-library/react'
import React, { FC, ReactElement, ReactNode } from 'react'
import Form from '../components/Form'
import mockErrors from '../mocks/mockErrors'

interface WrapperProps {
  theme?: typeof scwuiTheme
}

const Wrapper: FC<WrapperProps> = ({ theme = scwuiTheme, children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotWithPortal,
} = makeHelpers(Wrapper, { classNameReplacer: className => className })

export const shouldMatchEmotionSnapshotFormWrapper = (
  children: ReactNode,
  options?: {
    transform: (node: RenderResult) => Promise<void> | void
  },
) =>
  shouldMatchEmotionSnapshot(
    <Form errors={mockErrors}>{() => children}</Form>,
    options,
  )

export const renderWithWrapper = (children: ReactElement) =>
  render(children, {
    wrapper: () => (
      <ThemeProvider theme={scwuiTheme}>
        <Form errors={mockErrors}>{children}</Form>
      </ThemeProvider>
    ),
  })

export const mockRandom = () =>
  jest.spyOn(global.Math, 'random').mockReturnValue(0.4155913669444804)

export const restoreRandom = () =>
  jest.spyOn(global.Math, 'random').mockRestore()
