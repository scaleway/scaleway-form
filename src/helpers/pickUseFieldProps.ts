import { AnyObject } from 'final-form'
import { UseFieldConfig } from 'react-final-form'
import { BaseFieldProps } from '../types'

const pickUseFieldProps = <T = AnyObject, K = AnyObject>({
  afterSubmit,
  allowNull,
  beforeSubmit,
  data,
  defaultValue,
  format,
  formatOnBlur,
  initialValue,
  isEqual,
  multiple,
  parse,
  subscription,
  validate,
  validateFields,
  value,
}: BaseFieldProps<T, K>): UseFieldConfig<T, K> => ({
  afterSubmit,
  allowNull,
  beforeSubmit,
  data,
  defaultValue,
  format,
  formatOnBlur,
  initialValue,
  isEqual,
  multiple,
  parse,
  subscription,
  validate,
  validateFields,
  value,
})

export default pickUseFieldProps
