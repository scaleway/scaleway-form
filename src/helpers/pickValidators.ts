import type { ValidatorObject, ValidatorProps } from '../types'
import validators from '../validators'
import type { ValidatorFn } from '../validators/types'

export const pickValidators = <InputValue = unknown>(args: ValidatorProps) =>
  Object.entries(args)
    .map(([key, value]) =>
      value !== undefined
        ? (validators as Record<string, ValidatorFn>)[
            key as keyof typeof validators
          ]?.(value)
        : undefined,
    )
    .filter(validator => !!validator) as ValidatorObject<InputValue>[]
