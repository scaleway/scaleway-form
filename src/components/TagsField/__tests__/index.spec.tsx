import { TagsField } from '../..'
import { shouldMatchEmotionSnapshotFormWrapper } from '../../../../.jest/helpers'

describe('ToggleField', () => {
  test('should render correctly', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <TagsField name="test" placeholder="placeholder" />,
    ))

  test('should render correctly disabled', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <TagsField name="test-disabled" placeholder="placeholder" disabled />,
    ))

  test('should render correctly with default tags', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <TagsField
        name="test-tags"
        placeholder="placeholder"
        tags={['tags-1', 'tags-2']}
      />,
    ))
})
