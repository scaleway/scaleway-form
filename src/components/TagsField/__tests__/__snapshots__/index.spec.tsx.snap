// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchField should render correctly 1`] = `
<DocumentFragment>
  .cache-4own23 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 8px;
  border-radius: 4px;
  border: 1px solid #d4dae7;
}

.cache-4own23:focus-within {
  border: 1px solid #4f0599;
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-4own23>* {
  margin: 6px;
}

.cache-11neoz9 {
  font-size: 16px;
  color: #4a4f62;
  border: none;
  outline: none;
  background-color: #ffffff;
}

.cache-11neoz9::-webkit-input-placeholder {
  color: #b2b6c3;
}

.cache-11neoz9::-moz-placeholder {
  color: #b2b6c3;
}

.cache-11neoz9:-ms-input-placeholder {
  color: #b2b6c3;
}

.cache-11neoz9::placeholder {
  color: #b2b6c3;
}

<form
    novalidate=""
  >
    <div
      class="cache-4own23"
    >
      <input
        aria-label="test"
        class="cache-11neoz9"
        name="test"
        placeholder="placeholder"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`SwitchField should render correctly disabled 1`] = `
<DocumentFragment>
  .cache-4own23 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 8px;
  border-radius: 4px;
  border: 1px solid #d4dae7;
}

.cache-4own23:focus-within {
  border: 1px solid #4f0599;
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-4own23>* {
  margin: 6px;
}

<form
    novalidate=""
  >
    <div
      class="cache-4own23"
    />
  </form>
</DocumentFragment>
`;

exports[`SwitchField should render correctly with default tags 1`] = `
<DocumentFragment>
  .cache-4own23 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 8px;
  border-radius: 4px;
  border: 1px solid #d4dae7;
}

.cache-4own23:focus-within {
  border: 1px solid #4f0599;
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-4own23>* {
  margin: 6px;
}

.cache-11aaxx7-StyledContainer-base {
  border-radius: 4px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #f6f6f8;
  height: 24px;
  padding-left: 8px;
  padding-right: 8px;
}

.cache-1vwmh8m-StyledText {
  color: #4a4f62;
  font-size: 14px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  max-width: 350px;
  overflow: hidden;
  white-space: pre;
  text-overflow: ellipsis;
}

.cache-79dopx-StyledTouchable-StyledTouchable {
  border: 0;
  transition-property: opacity;
  transition-duration: 0.15s;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  padding: 0;
  margin: 0;
  margin-left: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.cache-79dopx-StyledTouchable-StyledTouchable[aria-disabled='true'] {
  opacity: 0.5;
  cursor: not-allowed;
}

.cache-79dopx-StyledTouchable-StyledTouchable[aria-disabled='false'] {
  cursor: pointer;
  touch-action: manipulation;
}

.cache-79dopx-StyledTouchable-StyledTouchable:hover,
.cache-79dopx-StyledTouchable-StyledTouchable:focus {
  background-color: #d7ccf0;
}

.cache-79dopx-StyledTouchable-StyledTouchable:hover svg,
.cache-79dopx-StyledTouchable-StyledTouchable:focus svg {
  fill: #390171;
}

.cache-1vsf6gj-StyledIcon-sizeStyles {
  fill: #b2b6c3;
  height: 16px;
  width: 16px;
  min-width: 16px;
  min-height: 16px;
}

.cache-1vsf6gj-StyledIcon-sizeStyles.cache-1vsf6gj-StyledIcon-sizeStyles {
  vertical-align: middle;
}

.cache-11neoz9 {
  font-size: 16px;
  color: #4a4f62;
  border: none;
  outline: none;
  background-color: #ffffff;
}

.cache-11neoz9::-webkit-input-placeholder {
  color: #b2b6c3;
}

.cache-11neoz9::-moz-placeholder {
  color: #b2b6c3;
}

.cache-11neoz9:-ms-input-placeholder {
  color: #b2b6c3;
}

.cache-11neoz9::placeholder {
  color: #b2b6c3;
}

<form
    novalidate=""
  >
    <div
      class="cache-4own23"
    >
      <div
        class="e1vt8sjg2 cache-11aaxx7-StyledContainer-base"
      >
        <span
          aria-disabled="false"
          class="cache-1vwmh8m-StyledText e1vt8sjg1"
        >
          tags-1
        </span>
        <button
          aria-disabled="false"
          aria-label="Close tag"
          class="e1vt8sjg0 e1kgqxfn0 cache-79dopx-StyledTouchable-StyledTouchable"
          type="button"
        >
          <svg
            class="sc-ui-icon etwatq50 cache-1vsf6gj-StyledIcon-sizeStyles"
            viewBox="0 0 24 24"
          >
            <path
              d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
            />
          </svg>
        </button>
      </div>
      <div
        class="e1vt8sjg2 cache-11aaxx7-StyledContainer-base"
      >
        <span
          aria-disabled="false"
          class="cache-1vwmh8m-StyledText e1vt8sjg1"
        >
          tags-2
        </span>
        <button
          aria-disabled="false"
          aria-label="Close tag"
          class="e1vt8sjg0 e1kgqxfn0 cache-79dopx-StyledTouchable-StyledTouchable"
          type="button"
        >
          <svg
            class="sc-ui-icon etwatq50 cache-1vsf6gj-StyledIcon-sizeStyles"
            viewBox="0 0 24 24"
          >
            <path
              d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
            />
          </svg>
        </button>
      </div>
      <input
        aria-label="test-tags"
        class="cache-11neoz9"
        name="test-tags"
        placeholder=""
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;
